### PR TITLE
Implement PHPStan testing

### PR DESCRIPTION
**Pull request by GwendolenLynch**
- Add as a dev requirement to composer.json
- Add configuration file to project root
- Update Travis CI to use

This one will be a bit of WiP to start, as the approach to the failing tests needs discussion.

### CODE REVIEW

File: src/User/composer.json [hunk 1]
```diff
@@ -17,13 +17,14 @@
         "exclude-from-classmap": ["/Tests/"]
     },
     "require": {
-        "msgphp/domain": "^1.0"
+        "msgphp/domain": "^1.0",
+        "msgphp/eav": "^1.0"
```
Comment (ro0NL) [inline src/User/composer.json]: this one is truly require-dev

File: src/User/Tests/Infra/Security/SecurityUserCheckerTest.php [hunk 2]
```diff
@@ -16,49 +16,44 @@
 
 final class SecurityUserCheckerTest extends TestCase
```
Comment (ro0NL) [inline src/User/Tests/Infra/Security/SecurityUserCheckerTest.php]: what happened here? :)

**Review by ro0NL — COMMENTED** [review]
not to familiar with phpstan yet.. but in general im :+1: for all the goodies :)

File: src/User/Tests/Infra/Security/SecurityUserCheckerTest.php [hunk 3]
```diff
@@ -16,49 +16,44 @@
...
 {
-    public function testPreAuth(): void
+    public function testPreAuthDisabledUser(): void
     {
+        $this->expectException(DisabledException::class);
```
Comment (ro0NL) [inline src/User/Tests/Infra/Security/SecurityUserCheckerTest.php]: in general i put expectException as late as possible, not causing false positives if i.e. User::__construct (below) throws such an exception

File: src/User/Tests/Infra/Security/SecurityUserCheckerTest.php [hunk 4]
```diff
@@ -16,49 +16,44 @@
 
 final class SecurityUserCheckerTest extends TestCase
 {
-    public function testPreAuth(): void
+    public function testPreAuthDisabledUser(): void
```
Comment (ro0NL) [inline src/User/Tests/Infra/Security/SecurityUserCheckerTest.php]: for consistency `testPreAuthWithDisabledUser` + `testPreAuthWithEnabledUser`

File: src/User/Tests/Infra/Security/SecurityUserCheckerTest.php [hunk 5]
```diff
@@ -16,49 +16,44 @@
...
+        $this->addToAssertionCount(1);
     }
 
     public function testPreAuthWithUnknownUser(): void
     {
-        $user = new User($this->getMockBuilder(UserIdInterface::class)->getMock(), 'foo@bar.baz', 'secret');
-
         $this->expectException(EntityNotFoundException::class);
 
+        $user = new User($this->getMockBuilder(UserIdInterface::class)->getMock(), 'foo@bar.baz', 'secret');
```
Comment (ro0NL) [inline src/User/Tests/Infra/Security/SecurityUserCheckerTest.php]: same.. i'd revert the order with expectException

File: src/User/phpstan.neon [hunk 6]
```diff
@@ -0,0 +1,13 @@
+parameters:
+
+    autoload_files:
+        - vendor/bin/.phpunit/phpunit-6.3/vendor/autoload.php
+
+    ignoreErrors:
+        # Ignored so that phpstan/phpstan-phpunit doesn't have to be installed
+        - '#PHPUnit_Framework_MockObject_MockObject given#'
+        # PHPDoc bug in SF
+        - '#Strict comparison using === between null and string will always evaluate to false#'
```
Comment (ro0NL) [inline src/User/phpstan.neon]: can it be more specific to `SF\Console::ask()`? not sure what the previous message was exactly ^^

File: src/User/Tests/Infra/Security/SecurityUserCheckerTest.php [hunk 7]
```diff
@@ -16,28 +16,26 @@
 
 final class SecurityUserCheckerTest extends TestCase
```
Comment (ro0NL) [inline src/User/Tests/Infra/Security/SecurityUserCheckerTest.php]: you can rebase btw.. this test is now gone :)

Comment (GawainLynch) [reply src/User/Tests/Infra/Security/SecurityUserCheckerTest.php]: Yup, I did about an hour or so ago after your last PR went in (a few new things popped up on the PHPStan tests)

**Comment by ro0NL** [comment]
@GawainLynch merge for now? fixing phpstan issues is a new story. Unless you have something going for travis... as mentioned using folded output for CC/CS/SA without failing is reasonable IMHO.

File: phpstan.neon [hunk 8]
```diff
@@ -0,0 +1,14 @@
+includes:
+    - vendor/phpstan/phpstan-phpunit/extension.neon
+    - vendor/phpstan/phpstan-phpunit/rules.neon
+    - vendor/phpstan/phpstan-phpunit/strictRules.neon
+
+parameters:
+    ignoreErrors:
+        # PHPDoc bug in SF
```
Comment (ro0NL) [inline phpstan.neon]: `# see https://github.com/symfony/symfony/pull/25644` + uncomment below

File: .travis.yml [hunk 9]
```diff
@@ -44,12 +45,13 @@ install:
...
   - bin/phpunit install
-  - if [[ $CS = true ]] ; then composer install --working-dir=.travis ; fi
+  - if [[ $CS = true ]] || [[ $SA = true ]] ; then composer install ; fi
 
 script:
   - if [[ ! $CC  ]] ; then bin/phpunit $PHPUNIT_ARGS ; fi
   - if [[ $CC = true ]] ; then bin/phpunit --coverage-clover=coverage.xml $PHPUNIT_ARGS ; fi
-  - if [[ $CS = true ]] ; then .travis/vendor/bin/php-cs-fixer fix --dry-run --verbose --diff src/ ; fi
+  - if [[ $SA = true ]] ; then vendor/bin/phpstan analyse --configuration phpstan.neon --level 7 src/ ; fi
+  - if [[ $CS = true ]] ; then vendor/bin/php-cs-fixer fix --dry-run --verbose --diff src/ ; fi
```
Comment (ro0NL) [inline .travis.yml]: reverse order? first php-cs-fixer then phpstan. or maybe if CS fails dont run SA anyway

**Review by ro0NL — APPROVED** [review]

**Pull request merged**